### PR TITLE
Clang-tidy modernize-loop-convert Framework

### PR DESCRIPTION
**Pull request by TTitcombe**
**Description of Work:**
This PR runs clang-tidy on Framework to use range-based for loops where possible

**To test:**
1. Code inspection
2. Tests pass

Note that if you search framework code for `for \(.*\.begin\(\);.*\.end\(\);.*\)` (regex for an old for loop) you will find some instances which remain. There are some instances where clang-tidy has decided they should not be / it doesn't know how to change the loops to range-based. These stragglers should be manually assessed at a later date

Fixes partially #22183

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/CalculateDynamicRange.cpp [hunk 1]
```diff
@@ -80,8 +80,12 @@ void CalculateDynamicRange::calculateQMinMax(MatrixWorkspace_sptr workspace,
   const auto &spectrumInfo = workspace->spectrumInfo();
   double min = std::numeric_limits<double>::max(),
          max = std::numeric_limits<double>::min();
+  // PARALLEL_FOR_NO_WSP_CHECK does not work with range-based for so NOLINT this
+  // block
   PARALLEL_FOR_NO_WSP_CHECK()
-  for (int64_t index = 0; index < static_cast<int64_t>(indices.size());
+  for (int64_t index = 0;
+       index <
+       static_cast<int64_t>(indices.size()); // NOLINT (modernize-for-loop)
```
Comment (DanNixon) [inline Framework/Algorithms/src/CalculateDynamicRange.cpp]: Can `NOLINT` be moved to the end of the `for` expression (next line down)?

File: Framework/Algorithms/src/ReflectometryWorkflowBase2.cpp [hunk 2]
```diff
@@ -726,15 +726,15 @@ ReflectometryWorkflowBase2::convertProcessingInstructionsToWorkspaceIndices(
   std::string converted = "";
   std::string currentNumber = "";
   std::string ignoreThese = "-,:+";
-  for (auto i = 0u; i < instructions.size(); ++i) {
-    if (std::find(ignoreThese.begin(), ignoreThese.end(), instructions[i]) !=
+  for (char instruction : instructions) {
```
Comment (DanNixon) [inline Framework/Algorithms/src/ReflectometryWorkflowBase2.cpp]: Can this be changed to `const char`?

File: Framework/Algorithms/src/ReflectometryWorkflowBase2.cpp [hunk 3]
```diff
@@ -757,15 +757,15 @@ ReflectometryWorkflowBase2::convertProcessingInstructionsToSpectrumNumbers(
...
```
Comment (DanNixon) [inline Framework/Algorithms/src/ReflectometryWorkflowBase2.cpp]: `const char`

File: Framework/DataHandling/src/LoadPSIMuonBin.cpp [hunk 4]
```diff
@@ -324,11 +324,10 @@ void LoadPSIMuonBin::generateUnknownAxis() {
   }
 
   // Create Errors
-  for (auto histogramIndex = 0u; histogramIndex < m_histograms.size();
-       ++histogramIndex) {
+  for (auto &histogram : m_histograms) {
```
Comment (DanNixon) [inline Framework/DataHandling/src/LoadPSIMuonBin.cpp]: Can this be `const auto`?

**Review by DanNixon — CHANGES_REQUESTED** [review]

File: Framework/Algorithms/src/CalculateDynamicRange.cpp (hunk 1 above)
Comment (TTitcombe) [reply Framework/Algorithms/src/CalculateDynamicRange.cpp]: Good spot, not sure why I put it in the middle of the for

**Review by DanNixon — APPROVED** [review]

**Pull request merged**